src/main/java/org/kuali/kfs/vnd/document/validation/impl/VendorRule.java

```diff
@@ -0,0 +1,1741 @@
+        // setup oldVendor convenience objects, make sure all possible sub-objects are populated
+        refreshSubObjects(oldVendor);
+
+        // setup newVendor convenience objects, make sure all possible sub-objects are populated
+        refreshSubObjects(newVendor);
+    }
+
+    /**
+     * Refreshes the references of vendor detail and its sub objects
+     *
+     * @param vendor VendorDetail document
+     */
+    protected void refreshSubObjects(VendorDetail vendor) {
+        if (vendor == null) {
+            return;
+        }
+
+        // If this is a division vendor, we need to do a refreshNonUpdateableReferences and also refreshes the vendor
+        // header, since the user aren't supposed to make any updates of vendor header's attributes while editing a
+        // division vendor
+        if (!vendor.isVendorParentIndicator()) {
+            vendor.refreshNonUpdateableReferences();
+            vendor.getVendorHeader().refreshNonUpdateableReferences();
+        } else {
+            // Retrieve the references objects of the vendor header of this vendor.
+            List<String> headerFieldNames = getObjectReferencesListFromBOClass(VendorHeader.class);
+            vendor.getVendorHeader().refreshNonUpdateableReferences();
+            SpringContext.getBean(PersistenceService.class).retrieveReferenceObjects(vendor.getVendorHeader(),
+                    headerFieldNames);
+
+            // We still need to retrieve all the other references of this vendor in addition to
+            // vendor header. Since this is a parent vendor, whose vendor header saving is handled manually,
+            // we have already retrieved references for vendor header's attributes above, so we should
+            // exclude retrieving reference objects of vendor header.
+            List<String> detailFieldNames = getObjectReferencesListFromBOClass(vendor.getClass());
+            detailFieldNames.remove(VendorConstants.VENDOR_HEADER_ATTR);
+            SpringContext.getBean(PersistenceService.class).retrieveReferenceObjects(vendor, detailFieldNames);
+        }
+
+        // refresh addresses
+        if (vendor.getVendorAddresses() != null) {
+            for (VendorAddress address : vendor.getVendorAddresses()) {
+                address.refreshNonUpdateableReferences();
+                if (address.getVendorDefaultAddresses() != null) {
+                    for (VendorDefaultAddress defaultAddress : address.getVendorDefaultAddresses()) {
+                        defaultAddress.refreshNonUpdateableReferences();
+                    }
+                }
+            }
+        }
+        // refresh contacts
+        if (vendor.getVendorContacts() != null) {
+            for (VendorContact contact : vendor.getVendorContacts()) {
+                contact.refreshNonUpdateableReferences();
+                // ==== CU Customization ====
+                if (contact.getVendorContactPhoneNumbers() != null) {
```

Comment:
should this use StringUtils.isNotEmpty instead?

Reply:
contact.getVendorContactPhoneNumbers() returns a collection, not a string. I used a non-null check here for consistency with the "vendor.getVendorContacts() != null" check from a few lines farther up.

Reply:
oh, thanks for the clarification